agent/util_test.go

```diff
@@ -119,3 +123,166 @@ func TestDurationFixer(t *testing.T) {
 	// Ensure we only processed the intended fieldnames
 	verify.Values(t, "", obj, expected)
 }
+
+// helperProcessSentinel is a sentinel value that is put as the first
+// argument following "--" and is used to determine if TestHelperProcess
+// should run.
+const helperProcessSentinel = "GO_WANT_HELPER_PROCESS"
+
+// helperProcess returns an *exec.Cmd that can be used to execute the
+// TestHelperProcess function below. This can be used to test multi-process
+// interactions.
+func helperProcess(s ...string) (*exec.Cmd, func()) {
+	cs := []string{"-test.run=TestHelperProcess", "--", helperProcessSentinel}
+	cs = append(cs, s...)
+
+	cmd := exec.Command(os.Args[0], cs...)
+	destroy := func() {
+		if p := cmd.Process; p != nil {
+			p.Kill()
+		}
+	}
+
+	return cmd, destroy
+}
+
+// This is not a real test. This is just a helper process kicked off by tests
+// using the helperProcess helper function.
+func TestHelperProcess(t *testing.T) {
+	args := os.Args
+	for len(args) > 0 {
+		if args[0] == "--" {
+			args = args[1:]
+			break
+		}
+
+		args = args[1:]
+	}
+
+	if len(args) == 0 || args[0] != helperProcessSentinel {
+		return
+	}
+
+	defer os.Exit(0)
+	args = args[1:] // strip sentinel value
+	cmd, args := args[0], args[1:]
+
+	switch cmd {
+	case "parent-signal":
+		// This subcommand forwards signals to a child process subcommand "print-signal".
+		testname := args[0]
+
+		limitProcessLifetime(2 * time.Minute)
+
+		cmd, destroy := helperProcess("print-signal")
+		defer destroy()
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+
+		if err := cmd.Start(); err != nil {
+			fmt.Fprintf(os.Stderr, "%s: child process failed to start: %v\n", testname, err)
+			os.Exit(1)
+		}
+
+		doneCh := make(chan struct{})
+		defer func() { close(doneCh) }()
+		logFn := func(err error) {
+			fmt.Fprintf(os.Stderr, "%s: could not forward signal: %s\n", testname, err)
+			os.Exit(1)
+		}
+		ForwardSignals(cmd, logFn, doneCh)
+
+		if err := cmd.Wait(); err != nil {
+			fmt.Fprintf(os.Stderr, "%s: unexpected error waiting for child: %v", testname, err)
+			os.Exit(1)
+		}
+
+	case "print-signal":
+		// This subcommand is instrumented to help verify signals are passed correctly.
+
+		limitProcessLifetime(2 * time.Minute)
+
+		ch := make(chan os.Signal, 10)
+		signal.Notify(ch)
+		defer signal.Stop(ch)
+
+		fmt.Fprintf(os.Stdout, "ready\n")
+
+		s := <-ch
+
+		fmt.Fprintf(os.Stdout, "signal: %s\n", s)
+
+	default:
+		fmt.Fprintf(os.Stderr, "Unknown command: %q\n", cmd)
+		os.Exit(2)
+	}
+}
+
+// limitProcessLifetime installs a background goroutine that self-exits after
+// the specified duration elapses to prevent leaking processes from tests that
+// may spawn them.
+func limitProcessLifetime(dur time.Duration) {
+	go time.AfterFunc(dur, func() {
+		os.Exit(99)
+	})
+}
+
+func TestForwardSignals(t *testing.T) {
+	for _, s := range forwardSignals {
+		testForwardSignal(t, s)
```

Comment:
I would wrap each one of these in a subtest `t.Run()` with the signal name/number.

Reply:
I totally forgot about `t.Run`. I'll also remove the `testname` log prefix stuff as it's not needed to understand the failures now.